File: app/src/main/java/de/bahnhoefe/deutschlands/bahnhofsfotos/model/HighScore.java
package de.bahnhoefe.deutschlands.bahnhofsfotos.model;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

public class HighScore {

    private final List<HighScoreItem> items = new ArrayList<>();

    public List<HighScoreItem> getItems() {
        return items;
    }


    public static class HighScoreDeserializer implements JsonDeserializer<HighScore> {
        public HighScore deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
                throws JsonParseException {
            var highScore = new HighScore();

            AtomicInteger position = new AtomicInteger(0);
            AtomicInteger lastPhotos = new AtomicInteger(0);
            json.getAsJsonObject().entrySet().stream()
                    .map(entry -> toHighScoreItem(position, lastPhotos, entry))
                    .forEach(highScore.items::add);

            return highScore;
        }

        private HighScoreItem toHighScoreItem(AtomicInteger position, AtomicInteger lastPhotos, Map.Entry<String, JsonElement> entry) {
            int photos = entry.getValue().getAsInt();
            if (lastPhotos.get() == 0 || lastPhotos.get() > photos) {
                position.incrementAndGet();
            }
            lastPhotos.set(photos);
            return new HighScoreItem(entry.getKey(), photos, position.get());
        }
    }

}

File: app/src/main/kotlin/de/bahnhoefe/deutschlands/bahnhofsfotos/model/HighScore.kt
package de.bahnhoefe.deutschlands.bahnhofsfotos.model

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import java.lang.reflect.Type
import java.util.concurrent.atomic.AtomicInteger

class HighScore {
    private val items: MutableList<HighScoreItem> = ArrayList()
    fun getItems(): List<HighScoreItem> {
        return items
    }

    class HighScoreDeserializer : JsonDeserializer<HighScore> {
        @Throws(JsonParseException::class)
        override fun deserialize(
            json: JsonElement,
            typeOfT: Type,
            context: JsonDeserializationContext
        ): HighScore {
            val highScore = HighScore()
            val position = AtomicInteger(0)
            val lastPhotos = AtomicInteger(0)
            json.asJsonObject.entrySet().stream()
                .map { entry: Map.Entry<String, JsonElement> ->
                    toHighScoreItem(
                        position,
                        lastPhotos,
                        entry
                    )
                }
                .forEach { e: HighScoreItem -> highScore.items.add(e) }
            return highScore
        }

        private fun toHighScoreItem(
            position: AtomicInteger,
            lastPhotos: AtomicInteger,
            entry: Map.Entry<String, JsonElement>
        ): HighScoreItem {
            val photos = entry.value.asInt
            if (lastPhotos.get() == 0 || lastPhotos.get() > photos) {
                position.incrementAndGet()
            }
            lastPhotos.set(photos)
            return HighScoreItem(entry.key, photos, position.get())
        }
    }
}